### PR TITLE
[BEAM-6097] NemoRunner

### PR DESCRIPTION
**Pull request by wonook**
This is a PR that adds the documentations for the Nemo Runner on the Beam website, as well as an appropriate profile to the maven artifact `pom.xml` file. At the moment, the Apache Nemo 0.1 release process is in progress, and I'm expecting the release process to be finished by this or next week hopefully. The maven `pom.xml` will be perfectly valid by then. As for now, in order to test the Nemo Runner, it has to be built on the local machine from source (get source from https://github.com/apache/incubator-nemo and run the command `man clean install -DskipTests`) and the version can be specified as `0.1-SNAPSHOT`. I've confirmed that the commands that I've entered in the documents work fine on my machine (Mac OS). I've mostly followed the format of other runners to provide the documentation for Apache Nemo Runner, but please let me know if there are additional changes that has to be made. Thanks!

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by wonook** [comment]
Run Spotless PreCommit

**Comment by wonook** [comment]
Run CommunityMetrics PreCommit

**Comment by wonook** [comment]
Run Python PreCommit

**Comment by wonook** [comment]
Run Go PreCommit

**Comment by wonook** [comment]
Run Python PreCommit

**Comment by wonook** [comment]
Run Python PreCommit

**Comment by wonook** [comment]
Run Go PreCommit

**Comment by mxm** [comment]
Run Python PreCommit

**Comment by wonook** [comment]
@mxm Thanks for the comment! We actually have been aware of the test suite and have set up an issue to check our Runner with the ValidatesRunner tests, but we haven't tried it out yet. 
Yet, we have run a variety of applications (including batch workloads like MR, MLR, ALS, and also the NexMark benchmark) which we were successful in running. The Runner will be continuously improved.

File: website/src/documentation/runners/nemo.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+After running `mvn package`, run `ls target` and you should see (when your artifactId is `beam-examples`
+and the version is `1.0.0`):
+
+```
+beam-examples-1.0.0-shaded.jar
+```
+
+With this shaded jar, you may use the `bin/run_beam.sh` shell script as following:
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
With this shaded jar, you can use the `bin/run_beam.sh` shell script as follows:
```

File: website/src/documentation/runners/nemo.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    -user_args "`pwd`/examples/resources/test_input_wordcount `pwd`/examples/resources/test_output_wordcount"
+```
+
+To use Nemo using YARN, set the `-deploy_mode` flag on Nemo to `yarn`. 
+
+More instructions can be seen on the README of the [Apache Nemo GitHub](https://github.com/apache/incubator-nemo).
+
+## Pipeline Options for the Nemo Runner
+
+When executing your pipeline with the Nemo Runner, you should consider the following pipeline options
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
When executing your pipeline with the Nemo Runner, you should consider the following pipeline options:
```

File: website/src/documentation/runners/nemo.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+| `maxBundleTimeMillis` | The maximum time to wait before finalizing a bundle (in milliseconds). |   1000 |
+
+More options are to be added to the list, to fully support the various options that Nemo supports.
+
+## Additional Information and Caveats
+
+### Using the Run_beam.sh script
+
+When submitting a Nemo application to the cluster, it is common to use the `bin/run_beam.sh` script that is
+provided within the Nemo installation. It also provides a richer set of options that you can pass on to
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
provided within the Nemo installation. The script also provides a richer set of options that you can pass on to
```

File: website/src/documentation/runners/nemo.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+More instructions can be seen on the README of the [Apache Nemo GitHub](https://github.com/apache/incubator-nemo).
+
+## Pipeline Options for the Nemo Runner
+
+When executing your pipeline with the Nemo Runner, you should consider the following pipeline options
+
+| Field       | Description           | Default Value  |
+| ------------- |---------------| -----:|
+| `runner`      | The pipeline runner to use. This option allows you to determine the pipeline runner at runtime. | Set to `NemoRunner` to run using Nemo |
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
| `runner`      | The pipeline runner to use. This option allows you to determine the pipeline runner at runtime. | Set to `NemoRunner` to run using Nemo. |
```

File: website/src/documentation/runners/nemo.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            <artifactId>slf4j-log4j12</artifactId>
+        </exclusion>
+    </exclusions>
+</dependency>
+```
+
+## Deploying Nemo with your Application
+
+A self-contained application could be easier to manage and fully use the functionalities that Nemo provides.
+Simply add the dependency shown above and shade the application jar using the maven shade plugin:
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
Simply add the dependency shown above and shade the application JAR using the Maven Shade plugin:
```

File: website/src/documentation/runners/nemo.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-log4j12</artifactId>
+        </exclusion>
+    </exclusions>
+</dependency>
+```
+
+## Deploying Nemo with your Application
+
+A self-contained application could be easier to manage and fully use the functionalities that Nemo provides.
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
A self-contained application might be easier to manage and allows you to fully use the functionality that Nemo provides.
```

File: website/src/documentation/runners/nemo.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+# Using the Apache Nemo Runner
+
+The Apache Nemo Runner can be used to execute Beam pipelines using [Apache Nemo](http://nemo.apache.org).
+The Nemo Runner can optimize Beam pipelines with the Nemo compiler through various optimization passes
+and execute them in distributed fashion using the Nemo runtime. Deploying a self-contained application
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
and execute them in a distributed fashion using the Nemo runtime. You can also deploy a self-contained application
```

File: website/src/documentation/runners/nemo.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+for local mode or running using resource managers like YARN or Mesos are possible.
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
for local mode or run using resource managers like YARN or Mesos.
```

File: website/src/documentation/runners/nemo.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+* Integration with YARN and other components of the Apache Hadoop ecosystem
+* Support for the various optimizations provided by the Nemo optimizer
+
+The [Beam Capability Matrix]({{ site.baseurl }}/documentation/runners/capability-matrix/) documents the
+supported capabilities of the Nemo Runner.
+
+## Nemo Runner prerequisites and setup
+
+The Nemo Runner can be used simply by adding a dependency on a version of the Nemo runner newer than `0.1`
+to your pom.xml as followings:
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
to your pom.xml as follows:
```

File: website/src/documentation/runners/nemo.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+          <transformer
+            implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+        </transformers>
+      </configuration>
+    </execution>
+  </executions>
+</plugin>
+```
+
+After running `mvn package`, run `ls target` and you should see (when your artifactId is `beam-examples`
```
Comment (melap) [inline website/src/documentation/runners/nemo.md]:
```suggestion
After running `mvn package`, run `ls target` and you should see the following output (in this example, your artifactId is `beam-examples`
```

**Comment by wonook** [comment]
@melap Lovely! I've applied your suggestions 😄 Thanks for catching out the details!

**Comment by wonook** [comment]
Run Portable_Python PreCommit

**Comment by wonook** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by wonook** [comment]
Run Portable_Python PreCommit

**Comment by kennknowles** [comment]
@mxm cross-posting that I took a minute to see if I could just plug in our old maven configs on https://github.com/apache/incubator-nemo/pull/182 but didn't get it over the finish line

**Comment by mxm** [comment]
Cool, thanks for looking into that. Running ValidatesRunner in the Nemo repo is useful. How about also integrating it in the Beam repo? It looks like that could be easier because we wouldn't have to deal with Maven.

**Comment by kennknowles** [comment]
@mxm could be good - it might be something to bring to the dev list about how to set it up, what tests to run, when, & why, greenness expectations, etc. WDYT?

**Comment by mxm** [comment]
I think it makes sense to discuss this on the dev list. We haven't fully figured out how to treat external Runners regarding CI tests.

**Comment by wonook** [comment]
Run Portable_Python PreCommit

**Comment by wonook** [comment]
Run Python PreCommit

**Comment by wonook** [comment]
Run Python PreCommit

**Comment by wonook** [comment]
Hi 😄 I must have missed out the notification regarding the ValidatesRunner and Nexmark. We've already been running benchmarks using Nexmark, and [nemo#182](https://github.com/apache/incubator-nemo/pull/182) is currently in progress. Please let me know if there are any further changes required for this PR! Thanks!

**Comment by mxm** [comment]
Run Java PreCommit

**Comment by mxm** [comment]
Unrelated flakiness of Elasticsearch tests.

Will merge this in a bit if there are no further comments.

**Comment by mxm** [comment]
@wonook Sorry, one more thing. Could you please squash the commit history into relevant commits?

**Comment by wonook** [comment]
@mxm no problem. I've squashed them 😄

**Pull request merged**